sklearn/gaussian_process/kernels.py

```diff
@@ -0,0 +1,1632 @@
+            Determines whether the gradient with respect to the kernel
+            hyperparameter is determined. Only supported when Y is None.
+
+        Returns
+        -------
+        K : array, shape (n_samples_X, n_samples_Y)
+            Kernel k(X, Y)
+
+        K_gradient : array (opt.), shape (n_samples_X, n_samples_X, n_dims)
+            The gradient of the kernel k(X, X) with respect to the
+            hyperparameter of the kernel. Only returned when eval_gradient
+            is True.
+        """
+        X = np.atleast_2d(X)
+        if Y is None:
+            dists = squareform(pdist(X, metric='sqeuclidean'))
+            tmp = dists / (2 * self.alpha * self.l ** 2)
+            base = (1 + tmp)
+            K = base ** -self.alpha
+            np.fill_diagonal(K, 1)
+        else:
+            if eval_gradient:
+                raise ValueError(
+                    "Gradient can only be evaluated when Y is None.")
+            dists = cdist(X, Y, metric='sqeuclidean')
+            K = (1 + dists / (2 * self.alpha * self.l ** 2)) ** -self.alpha
+
+        if eval_gradient:
+            # gradient with respect to l
+            if "l" in self.theta_vars:
+                l_gradient = dists * K / (self.l ** 2 * base)
+                l_gradient = l_gradient[:, :, np.newaxis]
+            else:  # l is kept fixed
+                l_gradient = np.empty((K.shape[0], K.shape[1], 0))
+
+            # gradient with respect to alpha
+            if "alpha" in self.theta_vars:
+                alpha_gradient = \
+                    K * (-self.alpha * np.log(base)
+                         + dists / (2 * self.l ** 2 * base))
+                alpha_gradient = alpha_gradient[:, :, np.newaxis]
+            else:  # alpha is kept fixed
+                alpha_gradient = np.empty((K.shape[0], K.shape[1], 0))
+
+            return K, np.dstack((l_gradient, alpha_gradient))
+        else:
+            return K
+
+    def __repr__(self):
+        return "{0}(alpha={1:.3g}, l={2:.3g})".format(
+            self.__class__.__name__, self.alpha, self.l)
+
+
+class ExpSineSquared(Kernel):
+    """Exp-Sine-Squared kernel.
+
+    The ExpSineSquared kernel allows modeling periodic functions. It is
+    parameterized by a length-scale parameter l>0 and a periodicity parameter
+    p>0. Only the isotropic variant where l is a scalar is supported at the
+    moment. The kernel given by:
+
+    k(x_i, x_j) =  exp(-2 sin(\pi / p * d(x_i, x_j)) / l)^2
+
+    Parameters
+    ----------
+    l : float > 0, default: 1.0
+        The length scale of the kernel.
+
+    p : float > 0, default: 1.0
+        The periodicity of the kernel.
+
+    l_bounds : pair of floats >= 0, default: (1e-5, 1e5)
+        The lower and upper bound on l
+
+    p_bounds : pair of floats >= 0, default: (1e-5, 1e5)
+        The lower and upper bound on p
+    """
+    def __init__(self, l=1.0, p=1.0, l_bounds=(1e-5, 1e5),
+                 p_bounds=(1e-5, 1e5)):
```

Comment:
Touples are fine, and in case someone wants to grid-search them (not sure if that is possible currently?) they make more sense. Though grid-searching these is probably a bad idea ^^